File: auth-data-postgres-client/src/main/java/com/cartobucket/auth/postgres/client/entities/mappers/ClientCodeMapper.java
/*
 * Copyright 2023 Bryce Groff (Cartobucket LLC)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
 * to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions
 * of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
 * BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package com.cartobucket.auth.postgres.client.entities.mappers;

import com.cartobucket.auth.data.domain.ClientCode;
import com.cartobucket.auth.data.domain.Scope;

import java.util.Collections;
import java.util.UUID;

public class ClientCodeMapper {
    public static ClientCode from (com.cartobucket.auth.postgres.client.entities.ClientCode clientCode, com.cartobucket.auth.data.services.ScopeService scopeService) {
        var _clientCode = new ClientCode();
        _clientCode.setCode(clientCode.getCode());
        _clientCode.setClientId(String.valueOf(clientCode.getClientId()));
        _clientCode.setCodeChallenge(clientCode.getCodeChallenge());
        _clientCode.setCodeChallengeMethod(clientCode.getCodeChallengeMethod());
        _clientCode.setId(clientCode.getId());
        _clientCode.setNonce(clientCode.getNonce());
        _clientCode.setState(clientCode.getState());
        _clientCode.setAuthorizationServerId(clientCode.getAuthorizationServerId());
        _clientCode.setRedirectUri(clientCode.getRedirectUri());
        _clientCode.setUserId(clientCode.getUserId());
        // Convert scope IDs back to Scope entities
        if (clientCode.getScopeIds() != null && !clientCode.getScopeIds().isEmpty()) {
            var scopes = clientCode.getScopeIds().stream()
                    .filter(scopeId -> scopeId != null) // Filter out null scope IDs
                    .map(scopeId -> {
                        try {
                            return scopeService.getScope(scopeId);
                        } catch (Exception e) {
                            // Log the error but don't fail the entire operation
                            System.err.println("Failed to load scope with ID: " + scopeId + " - " + e.getMessage());
                            return null;
                        }
                    })
                    .filter(scope -> scope != null) // Filter out failed scope lookups
                    .toList();
            _clientCode.setScopes(scopes);
        } else {
            _clientCode.setScopes(Collections.emptyList());
        }
        _clientCode.setCreatedOn(clientCode.getCreatedOn());
        return _clientCode;
    }

    public static com.cartobucket.auth.postgres.client.entities.ClientCode to (ClientCode clientCode) {
        var _clientCode = new com.cartobucket.auth.postgres.client.entities.ClientCode();
        _clientCode.setCode(clientCode.getCode());
        _clientCode.setClientId(UUID.fromString(clientCode.getClientId()));
        _clientCode.setCodeChallenge(clientCode.getCodeChallenge());
        _clientCode.setCodeChallengeMethod(clientCode.getCodeChallengeMethod());
        _clientCode.setId(clientCode.getId());
        _clientCode.setNonce(clientCode.getNonce());
        _clientCode.setState(clientCode.getState());
        _clientCode.setAuthorizationServerId(clientCode.getAuthorizationServerId());
        _clientCode.setRedirectUri(clientCode.getRedirectUri());
        _clientCode.setUserId(clientCode.getUserId());
        if (clientCode.getScopes() != null) {
            _clientCode.setScopeIds(clientCode.getScopes().stream()
                    .map(Scope::getId)
                    .toList());
        } else {
            _clientCode.setScopeIds(Collections.emptyList());
        }
        _clientCode.setCreatedOn(clientCode.getCreatedOn());
        return _clientCode;
    }
}

File: auth-data-postgres-client/src/main/kotlin/com/cartobucket/auth/postgres/client/entities/mappers/ClientCodeMapper.kt
/*
 * Copyright 2023 Bryce Groff (Cartobucket LLC)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
 * to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions
 * of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
 * BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package com.cartobucket.auth.postgres.client.entities.mappers

import com.cartobucket.auth.data.domain.ClientCode
import com.cartobucket.auth.data.services.ScopeService

object ClientCodeMapper {
    fun from(
        clientCode: com.cartobucket.auth.postgres.client.entities.ClientCode,
        scopeService: ScopeService,
    ): ClientCode {
        val domainClientCode = ClientCode()
        domainClientCode.code = clientCode.code
        domainClientCode.clientId = clientCode.clientId.toString()
        domainClientCode.codeChallenge = clientCode.codeChallenge
        domainClientCode.codeChallengeMethod = clientCode.codeChallengeMethod
        domainClientCode.id = clientCode.id
        domainClientCode.nonce = clientCode.nonce
        domainClientCode.state = clientCode.state
        domainClientCode.authorizationServerId = clientCode.authorizationServerId
        domainClientCode.redirectUri = clientCode.redirectUri
        domainClientCode.userId = clientCode.userId
        domainClientCode.createdOn = clientCode.createdOn

        // Get scopes using the service
        domainClientCode.scopes = clientCode.scopeIds?.let { scopeIds ->
            scopeService.getScopes(listOf(clientCode.authorizationServerId!!), scopeIds)
        } ?: emptyList()

        return domainClientCode
    }

    fun to(clientCode: ClientCode): com.cartobucket.auth.postgres.client.entities.ClientCode {
        val entityClientCode =
            com.cartobucket.auth.postgres.client.entities
                .ClientCode()
        entityClientCode.code = clientCode.code
        entityClientCode.clientId = clientCode.clientId?.let { java.util.UUID.fromString(it) }
        entityClientCode.codeChallenge = clientCode.codeChallenge
        entityClientCode.codeChallengeMethod = clientCode.codeChallengeMethod
        entityClientCode.id = clientCode.id
        entityClientCode.nonce = clientCode.nonce
        entityClientCode.state = clientCode.state
        entityClientCode.authorizationServerId = clientCode.authorizationServerId
        entityClientCode.redirectUri = clientCode.redirectUri
        entityClientCode.userId = clientCode.userId
        entityClientCode.createdOn = clientCode.createdOn
        entityClientCode.scopeIds = clientCode.scopes?.map { it.id!! }
        return entityClientCode
    }
}
